src/blk/kernel/KernelDevice.cc

```diff
@@ -562,21 +562,17 @@ void KernelDevice::_discard_stop()
   // Signal threads to stop, then wait for them to join
   {
     std::unique_lock l(discard_lock);
-    while (discard_threads.empty()) {
-      discard_cond.wait(l);
-    }
 
     for(auto &t : discard_threads) {
```

Comment:
With your new thread termination condition in _discard_thread():
_if (thr->stop && !discard_threads.empty())
        break;_

one might face a situation when _discard_stop() stays with a single running thread to drain the discard queue. Which could e.g. severely delay OSD's shutdown.
To fix that you might want to do the following here:
1. swap discard_threads content with a local empty container. Hence implicitly stopping new discards from being queued.
2. discard_drain().
3. stop/detach all the threads from the initial thread set.

Reply:
I don't think that's true - we're holding the discard lock the whole time here and when evaluating that condition in `_discard_thread()`.

Reply:
Indeed. My bad..